### PR TITLE
Added NewtonFool Attack

### PR DESCRIPTION
**Pull request by bveliqi**

### CODE REVIEW

File: foolbox/attacks/newtonfool.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+from .base import Attack
+from .base import call_decorator
+from ..utils import softmax
+
+import numpy as np
+
+
+class NewtonFoolAttack(Attack):
+    """Implements the NewtonFool Attacl.
```
Comment (jonasrauber) [inline foolbox/attacks/newtonfool.py]: typo

File: foolbox/attacks/newtonfool.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    The attack was introduced in [1]_.
+
+    References
+    ----------
+    .. [1] Uyeong Jang et al., "Objective Metrics and Gradient Descent Algorithms
+                                for Adversarial Examples in Machine Learning",
+        https://andrewxiwu.github.io/public/papers/2017/JWJ17-objective-metrics-and-gradient-descent-based-algorithms-for-adversarial-examples-in-machine-learning.pdf
+   """
```
Comment (jonasrauber) [inline foolbox/attacks/newtonfool.py]: indentation

File: foolbox/attacks/newtonfool.py (hunk 1 above)
Comment (bveliqi) [reply foolbox/attacks/newtonfool.py]: fixed.

File: foolbox/attacks/newtonfool.py (hunk 2 above)
Comment (bveliqi) [reply foolbox/attacks/newtonfool.py]: fixed.

File: foolbox/attacks/newtonfool.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+
+class NewtonFoolAttack(Attack):
+    """Implements the NewtonFool Attacl.
+
+    The attack was introduced in [1]_.
+
+    References
+    ----------
+    .. [1] Uyeong Jang et al., "Objective Metrics and Gradient Descent Algorithms
```
[the remark this answers is not in the document]
Comment (bveliqi) [reply foolbox/attacks/newtonfool.py]: done.

**Comment by bveliqi** [comment]
Thx for the feedback. Fixed, what you've mentioned and added the attack to the documentation.

**Comment by jonasrauber** [comment]
Great! One question: shouldn't this attack also work for targeted adversarials?

**Comment by wielandbrendel** [comment]
No, I looked into it with @bveliqi and I don't see any easy way to deploy this as a targeted attack. Can we merge this?

**Comment by jonasrauber** [comment]
it's on my agenda, but not yet

**Comment by wielandbrendel** [comment]
@jonasrauber Any update here? Would be great to merge.

File: foolbox/attacks/newtonfool.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        del unpack
+
+        if not a.has_gradient():
+            return
+
+        if a.target_class() is not None:
+            logging.fatal('NewtonFool is an untargeted adversarial attack.')
+            return
+
+        reshape = np.reshape(a.original_image, [-1])
```
Comment (jonasrauber) [inline foolbox/attacks/newtonfool.py]: this reshape is not needed; just pass `a.original_image` to np.linalg.norm

File: foolbox/attacks/newtonfool.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            # instead of using the logits and the gradient of the logits,
+            # we use a numerically stable implementation of the cross-entropy
+            # and expect that the deep learning frameworks also use such a
+            # stable implemenation to calculate the gradient
+            # grad is calculated from CE but we want softmax
+            # -> revert chain rule
+            gradients = - gradients / score
+
+            # (2) calculate gradient norm
+            gradient_l2_norm = np.linalg.norm(np.reshape(gradients, [-1]))
```
Comment (jonasrauber) [inline foolbox/attacks/newtonfool.py]: this reshape is not needed; just pass `gradients` to np.linalg.norm

File: foolbox/attacks/newtonfool.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    @staticmethod
+    def _delta(eta, norm, score, gradient_norm, num_classes):
+        a = eta * norm * gradient_norm
+        b = score - 1.0 / num_classes
+        return min(a, b)
+
+    @staticmethod
+    def _perturbation(delta, gradients, gradient_norm):
+        a = delta * gradients
+        b = float(gradient_norm ** 2)
```
Comment (jonasrauber) [inline foolbox/attacks/newtonfool.py]: why do we convert to float here?

File: foolbox/attacks/newtonfool.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                return
+
+            score = np.max(softmax(logits))
+            # instead of using the logits and the gradient of the logits,
+            # we use a numerically stable implementation of the cross-entropy
+            # and expect that the deep learning frameworks also use such a
+            # stable implemenation to calculate the gradient
+            # grad is calculated from CE but we want softmax
+            # -> revert chain rule
+            gradients = - gradients / score
```
Comment (jonasrauber) [inline foolbox/attacks/newtonfool.py]: nitpick: no space after that minus sign ;-)

File: foolbox/attacks/newtonfool.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    def _delta(eta, norm, score, gradient_norm, num_classes):
+        a = eta * norm * gradient_norm
+        b = score - 1.0 / num_classes
+        return min(a, b)
+
+    @staticmethod
+    def _perturbation(delta, gradients, gradient_norm):
+        a = delta * gradients
+        b = float(gradient_norm ** 2)
+        direction = -(a / b)
```
Comment (jonasrauber) [inline foolbox/attacks/newtonfool.py]: nitpick: unless there are numerical reasons, you could safe one matrix-scalar product by first calculating (delta / b) and then multiplying with gradients

**Comment by wielandbrendel** [comment]
Thanks for reviewing Behar's PR! Quick comment about last nitpick (I don't know how to add a comment within your review directly): these are all scalar x vector products if I am not mistaken, so this should not be an issue.

**Comment by jonasrauber** [comment]
@wielandbrendel `delta` and `b` are scalars, `gradients` is a vector (or matrix), if I am not mistaken ;-)

**Comment by wielandbrendel** [comment]
@jonasrauber You are right, my brain just completed "matrix-... product" to "matrix-vector product" and your "-scalar" went unnoticed ;-).

**Comment by bveliqi** [comment]
Okay, changed according your comments & merged with current master.

**Comment by jonasrauber** [comment]
I think you accidentally removed a minus sign.

**Comment by bveliqi** [comment]
Yes, your're right. Re-added it.

**Comment by jonasrauber** [comment]
There is still a typo, see failing tests
